pacrunner.go

```diff
@@ -202,17 +200,52 @@ func myIpAddress(call otto.FunctionCall) otto.Value {
 	return toValue("127.0.0.1")
 }
 
+func myIpAddressEx(_ otto.FunctionCall) otto.Value {
+	// https://chromium.googlesource.com/chromium/src/+/ee43fa5328856129f46566b2ea1be5811739681c/net/docs/proxy.md#resolving-client_s-ip-address-within-a-pac-script-using-myipaddressex
+	public := []string{"8.8.8.8", "2001:4860:4860::8888"}
+	if ips := probeRoutes(public); ips != "" {
+		return toValue(ips)
+	}
+	if ips := resolveHostname(true); len(ips) > 0 {
+		var b strings.Builder
+		for _, ip := range ips {
+			b.WriteRune(';')
+			b.WriteString(ip.String())
+		}
+		return toValue(strings.TrimPrefix(b.String(), ";"))
+	}
+	private := []string{"10.0.0.0", "172.16.0.0", "192.168.0.0", "FC00::"}
+	ips := probeRoutes(private)
+	return toValue(ips)
+}
+
+func probeRoutes(addresses []string) string {
+	var slice []string
+	set := map[string]struct{}{}
+	for _, address := range addresses {
+		localAddr := probeRoute(address)
+		if localAddr == "" {
+			continue
+		}
+		if _, ok := set[localAddr]; ok {
+			continue
+		}
+		set[localAddr] = struct{}{}
+		slice = append(slice, localAddr)
+	}
+	return strings.Join(slice, ";")
```

Comment:
If the order of the output does not need to match the order of the input, you don't really need the slice and you could simplify this:

```
    set := map[string]struct{}{}
    for _, address := range addresses {
        if localAddr := probeRoute(address); localAddr != "" {
            set[localAddr] = struct{}{}
        }
    }
    return strings.Join(slices.Collect(maps.Keys(set)), ";")
```

This does need Go 1.23 for `slices.Collect()` and `maps.Keys()` though.

Reply:
I was still on 1.22 but yes this would be nicer. I still need to check but I'll do it this way if I can confirm that order isn't important.

Reply:
I haven't found out whether order is important or not, but I think I'll just leave this as is, just in case it is.